realsense2_camera/include/mavros_timestamp.h

```diff
@@ -0,0 +1,432 @@
+    std::lock_guard<std::mutex> lg(mutex_);
+
+    trigger_sequence_offset_ = 0;
+    delay_pub_ = nh_.advertise<geometry_msgs::PointStamped>(kDelayTopic, 100);
+    state_ = ts_not_initalized;
+    framerate_ = framerate; // approximate framerate to do outlier checking
+    callback_ = callback;
+    static_time_shift_ = static_time_shift; // constant shift for bus transfers, sync-signal weirdness etc.
+    // Values for realsense infra channel:
+    // +10*1e-3 for 640x480, -6*1e-3 for 1280x720
+
+    cam_imu_sub_ =
+        nh_.subscribe(kCamImuSyncTopic, 100,
+                      &MavrosTrigger::camImuStampCallback, this);
+
+    ROS_INFO_STREAM(kLogPrefix << " Initialized with framerate " << framerate <<
+    " hz, static time offset "<< static_time_shift_<< " and subscribed to cam_imu_sub");
+  }
+
+  void start() {
+    std::lock_guard<std::mutex> lg(mutex_);
+
+    if (state_ != ts_not_initalized) {
+      //already started, nothing to do
+      return;
+    }
+
+    ROS_INFO_STREAM(kLogPrefix << " Started triggering.");
+
+    for (t_channel_id id : channel_set_) {
+      cache_queue_[id].clear();
+    }
+
+    trigger_sequence_offset_ = 0;
+
+    if (ros::service::exists(kTriggerService, false)) {
+      mavros_msgs::CommandTriggerControl req;
+      req.request.trigger_enable = true;
+      // This is NOT integration time, this is actually the sequence reset.
+      req.request.integration_time = 1.0;
+
+      ros::service::call(kTriggerService, req);
+
+      ROS_DEBUG("Called mavros trigger service! Success? %d Result? %d",
+               req.response.success, req.response.result);
+      state_ = ts_wait_for_sync;
+    } else {
+      ROS_ERROR("Mavros service not available!");
+      state_ = ts_disabled;
+    }
+
+  }
+
+  void waitMavros(){
+    ROS_INFO_STREAM("Waiting for mavros service....");
+    if(ros::service::waitForService(kTriggerService, ros::Duration(10))){
+      ROS_INFO_STREAM("... mavros service ready.");
+    } else {
+      ROS_ERROR_STREAM("... mavros service wait timeout.");
+      // service disable is handled in start() function
+    }
+  }
+
+  /*
+   * Adds a timestamp and sequence number of the camera without caching/publishing the image
+   *  - Keeps sync alive if images are not processed (e.g. because not needed)
+   */
+  void addCameraSequence(const t_channel_id &channel, const uint32_t camera_seq, const ros::Time &camera_stamp){
+    std::shared_ptr<t_cache> empty = {};
+    addCameraFrame(channel, camera_seq, camera_stamp, empty, 0.0);
+  }
+
+  /*
+   * Adds a timestamp, sequence number and image frame to be published by the callback once it's timing is corrected.
+   *  - Caches t_cache frame pointer until corresponding trigger time message from mavros is received
+   */
+  void addCameraFrame(const t_channel_id &channel, const uint32_t camera_seq,
+                      const ros::Time &camera_stamp, const std::shared_ptr<t_cache>& frame, const double exposure){
+    ros::spinOnce();
```

Comment:
why the spin?

Reply:
- added comment : "
    // allow threading subsystem to check for new IMU messages right before we lock.
"